### ISSUE TITLE
Health-based recloaking causes "pop-in / pop-out" behavior

### ISSUE THREAD

**Issue by tehhowch** [endless-sky/endless-sky#2836]
Reported on Steam: https://steamcommunity.com/app/404410/discussions/1/1458455461497401563/#c1471966894862283146

Video of the cloak/decloak/cloak loop.
https://youtu.be/IL0KMPFwp7k

Since the Archon (Cloaked) has free cloaking, it does not care about distance to target when deciding to de-cloak.
```
if(ship.Attributes().Get("cloak"))
	{
		// Never cloak if it will cause you to be stranded.
		if(ship.Attributes().Get("cloaking fuel") && !ship.Attributes().Get("ramscoop"))
		{
			double fuel = ship.Fuel() * ship.Attributes().Get("fuel capacity");
			fuel -= ship.Attributes().Get("cloaking fuel");
			if(fuel < ship.JumpFuel())
				return;
		}
		// Otherwise, always cloak if you are in imminent danger.
		static const double MAX_RANGE = 10000.;
		double nearestEnemy = MAX_RANGE;
		for(const auto &other : ships)
			if(other->GetSystem() == ship.GetSystem() && other->IsTargetable()
					&& other->GetGovernment()->IsEnemy(ship.GetGovernment())
					&& !other->IsDisabled())
				nearestEnemy = min(nearestEnemy,
					ship.Position().Distance(other->Position()));
		
		// If this ship has started cloaking, it must get at least 40% repaired
		// or 40% farther away before it begins decloaking again.
		double hysteresis = ship.Cloaking() ? 1.4 : 1.;
		double cloakIsFree = !ship.Attributes().Get("cloaking fuel");
		if(ship.Hull() + .5 * ship.Shields() < hysteresis
				&& (cloakIsFree || nearestEnemy < 2000. * hysteresis))
			command |= Command::CLOAK;
		
		// Also cloak if there are no enemies nearby and cloaking does
		// not cost you fuel.
		if(nearestEnemy == MAX_RANGE && cloakIsFree && !ship.GetTargetShip())
			command |= Command::CLOAK;
	}
```
For starters, the hysteresis value seems wrong - 40% repaired can be interpreted in a number of ways, but right now a value of 1.4 (out of 1.5 maximum) means a lot more than 40% repaired! 40% of 1.5 is 0.6. It is true that 1.4 is 0.4 more than the value of 1, though on the scale with 1.5 as 100% repaired, this is only 26% more repaired (going from 66% total HP to 93.33% total HP).

Second, if the AI-controlled target is reduced below the hysteresis threshold before it can finish decloaking, it immediately decides to recloak - as Ship::Cloaking() returns the current cloak value and not just 0 / 1, so the hysteresis value remains 1.4 until the ship has a cloak value of 0.

I suggest we add a threshold value such that the ship is always able to decloak fully after repairing to the threshold value.

**Comment by tehhowch** [comment]
I've addressed the pop-in/out behavior with #2837

I leave the selection of appropriate hysteresis values to another PR.

**Comment by endless-sky** [comment]
Addressed by merging in #2837.
